gh-113055: Use pointer for interp->obmalloc state.

nascheme opened the pull request:
For interpreters that share state with the main interpreter, this points to the same static memory structure.  For interpreters with their own obmalloc state, it is heap allocated.  Add free_obmalloc_arenas() which will free the obmalloc arenas and radix tree structures for interpreters with their own obmalloc state.

Note that the free is only done if obmalloc used blocks is zero.  If some blocks are used, it's not safe to free the memory since some extension might still be using it.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-113055 -->
* Issue: gh-113055
<!-- /gh-issue-number -->

nascheme committed:
Use pointer for interp->obmalloc state.

For interpreters that share state with the main interpreter, this points
to the same static memory structure.  For interpreters with their own
obmalloc state, it is heap allocated.  Add free_obmalloc_arenas() which
will free the obmalloc arenas and radix tree structures for interpreters
with their own obmalloc state.

nascheme committed:
Update Tools/c-analyzer/cpython/ignored.tsv.

The obmalloc_state_main global is okay and should be ignored.

ericsnowcurrently commented on Include/internal/pycore_runtime_init.h:
```diff
@@ -151,7 +151,8 @@ extern PyTypeObject _PyExc_MemoryError;
         .imports = IMPORTS_INIT, \
-        .obmalloc = _obmalloc_state_INIT(INTERP.obmalloc), \
+        /* initialized by _PyMem_init_obmalloc() */ \
+        .obmalloc = 0, \
```
Setting it to NULL is redundant.

```suggestion
```

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -2612,7 +2633,6 @@ _PyObject_DebugDumpAddress(const void *p)
     _PyMem_DumpTraceback(fileno(stderr), p);
 }
 
-
```
intentional?

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -606,6 +606,10 @@ init_interp_create_gil(PyThreadState *tstate, int gil)
 
 
+// defined in obmalloc.c
+int _PyMem_init_obmalloc(PyInterpreterState *interp, _PyRuntimeState *runtime);
```
```suggestion
extern int _PyMem_init_obmalloc(PyInterpreterState *interp, _PyRuntimeState *runtime);
```

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -2663,9 +2683,71 @@ _PyDebugAllocatorStats(FILE *out,
+        interp->obmalloc->initialized = true;
+    }
+#endif /* WITH_PYMALLOC */
+    return 1;
```
This feels more like a place we'd return -1 for "error" rather than a boolean for success.

```suggestion
            return -1;
        }
        interp->obmalloc->heap_allocated = true;
    }
    if (!interp->obmalloc->initialized) {
        // initialize the obmalloc->pools structure.  This must be done
        // before the obmalloc alloc/free functions can be called.
        poolp temp[OBMALLOC_USED_POOLS_SIZE] =
            _obmalloc_pools_INIT(interp->obmalloc->pools);
        memcpy(&interp->obmalloc->pools.used, temp, sizeof(temp));
        interp->obmalloc->initialized = true;
    }
#endif /* WITH_PYMALLOC */
    return 0;
```

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -2663,9 +2683,71 @@ _PyDebugAllocatorStats(FILE *out,
 }
 
+int _PyMem_init_obmalloc(PyInterpreterState *interp, _PyRuntimeState *runtime)
+{
```
Passing `runtime` isn't necessary, since it is accessible from the interpreter state.  (Plus, I don't think you need it in this function.)

```suggestion
int _PyMem_init_obmalloc(PyInterpreterState *interp)
{
```

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -2663,9 +2683,71 @@ _PyDebugAllocatorStats(FILE *out,
+            || (interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
+        interp->obmalloc = &obmalloc_state_main;
+        interp->obmalloc->heap_allocated = false;
+    } else {
```
```suggestion
    if (_Py_IsMainInterpreter(interp)
            || _PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        interp->obmalloc = &obmalloc_state_main;
        interp->obmalloc->heap_allocated = false;
    }
    else {
```

or keep the coupling as close as it was before:

```suggestion
    if (_Py_IsMainInterpreter(interp) {
        interp->obmalloc = &obmalloc_state_main;
        interp->obmalloc->heap_allocated = false;
    }
    else if (_PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        interp->obmalloc = _PyInterpreterState_Main()->obmalloc;
        assert(interp->obmalloc != NULL);
    }
    else {
```

or if you set `get_state()` back to the way it was (and return 0 for success):

```suggestion
    if (_Py_IsMainInterpreter(interp) {
        interp->obmalloc = &obmalloc_state_main;
        interp->obmalloc->heap_allocated = false;
    }
    else if (_PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        assert(interp->obmalloc == NULL);
        // It was already initialized for the main interpreter.
        return 0;
    }
    else {
```

ericsnowcurrently commented on Include/internal/pycore_obmalloc.h:
```diff
@@ -668,6 +668,15 @@ struct _obmalloc_state {
+    // before the malloc/free functions of obmalloc are called and must be
+    // done only once per obmalloc state.  The function _PyMem_init_obmalloc()
+    // does the initialization.
+    bool initialized;
```
This is only needed for `obmalloc_state_main` so we not add `int obmalloc_state_main_initialized = 0;` right next to it, and then drop this field?

(If we statically initialized `obmalloc_state_main` then we wouldn't need this at all.)

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -967,6 +968,12 @@ static int running_on_valgrind = -1;
 
+/* obmalloc state for main interpreter and shared by all interpreters without
+ * their own obmalloc state.  By not explicitly initalizing this structure, it
+ * will be allocated in the BSS which is a small performance win.  The radix
```
How much of a difference is it?  IIRC, we used a static initializer in 3.11.

I ask because things get a little bit simpler if we statically initialize.

Also, doesn't the common case in `_PyMem_init_obmalloc()` get faster since we can skip the initialize step (e.g. the `memcpy()`)?

ericsnowcurrently commented on Include/internal/pycore_obmalloc.h:
```diff
@@ -668,6 +668,15 @@ struct _obmalloc_state {
+    // true if this structure is heap allocated, by PyMem_RawCalloc().  For
+    // the main interpreter, this structure is statically allocated (in the
+    // BSS).  Using the BSS gives some performance win.
+    bool heap_allocated;
```
We don't need this since we can check `interp->obmalloc == &obmalloc_state_main`, right?

nascheme replied:
Yeah, that works.  I added a small helper function since `obmalloc_state_main` is local to obmalloc.c.

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -2663,9 +2683,71 @@ _PyDebugAllocatorStats(FILE *out,
+        // initialize the obmalloc->pools structure.  This must be done
+        // before the obmalloc alloc/free functions can be called.
+        poolp temp[OBMALLOC_USED_POOLS_SIZE] =
+            _obmalloc_pools_INIT(interp->obmalloc->pools);
```
Would it make a difference to make this static?

nascheme replied:
I don't see how that helps but maybe I miss the point.

ericsnowcurrently replied:
It was just to avoid the assignment each time.  I'm sure it doesn't matter much.

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -979,10 +986,8 @@ static inline OMState *
 get_state(void)
 {
     PyInterpreterState *interp = _PyInterpreterState_GET();
-    if (!has_own_state(interp)) {
-        interp = _PyInterpreterState_Main();
-    }
-    return &interp->obmalloc;
+    assert(interp->obmalloc != NULL); // otherwise not initialized or freed
+    return interp->obmalloc;
```
The existing code is still correct, isn't it?

```suggestion
    if (!has_own_state(interp)) {
        assert(interp->obmalloc == NULL);
        interp = _PyInterpreterState_Main();
    }
    assert(interp->obmalloc != NULL); // otherwise not initialized or freed
    return interp->obmalloc;
```

Alternately, we don't need to bother setting `interp->obmalloc` for the main interpreter:

```suggestion
    if (_Py_IsMainInterpreter(interp)) {
        assert(interp->obmalloc == NULL);
        return &obmalloc_state_main;
    }
    else if (_PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        assert(interp->obmalloc == NULL);
        return &obmalloc_state_main;
    }
    assert(interp->obmalloc != NULL); // otherwise not initialized or freed
    return interp->obmalloc;
```

Is the intention to avoid the branch for a little performance?  If so, how much does it win us?  I'd expect the branch predictor to help out somewhat.

nascheme replied:
Yes, the idea is to avoid branches.  The difference is probably tiny but this is a quite a hot path.

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -1060,10 +1071,20 @@ _PyInterpreterState_FinalizeAllocatedBlocks(PyInterpreterState *interp)
+            // should be warning when this happens.  It should be possible to
+            // use a tool like "-fsanitize=address" to track down these leaks.
+            free_obmalloc_arenas(interp);
+        }
```
I was going to say it makes me uncomfortable doing this here.  However, after 2 seconds it actually made sense to me that we'd free the arenas in a function called `_PyInterpreterState_FinalizeAllocatedBlocks()`. 😄

ericsnowcurrently commented on Objects/obmalloc.c:
```diff
@@ -2663,9 +2683,71 @@ _PyDebugAllocatorStats(FILE *out,
+            return 0;
+        }
+        interp->obmalloc->heap_allocated = true;
+    }
```
Alternately, we don't actually need to set `interp->obmalloc` for the main interpreter (and returning 0 for success):

```suggestion
    struct _obmalloc_state *state;
    if (_Py_IsMainInterpreter(interp)) {
        assert(interp->obmalloc == NULL);
        state = &obmalloc_state_main;
        state->heap_allocated = false;
    }
    else if (_PyInterpreterState_HasFeature(interp, Py_RTFLAGS_USE_MAIN_OBMALLOC)) {
        assert(interp->obmalloc == NULL);
        // It was already initialized for the main interpreter.
        return 0;
    }
    else {
        state = PyMem_RawCalloc(1, sizeof(struct _obmalloc_state));
        if (state == NULL) {
            return -1;
        }
        state->heap_allocated = true;
        interp->obmalloc = state;
    }
```

ericsnowcurrently commented on Include/internal/pycore_interp.h:
```diff
@@ -166,7 +166,17 @@ struct _is {
     struct atexit_state atexit;
 
-    struct _obmalloc_state obmalloc;
+    // Per-interpreter state for the obmalloc allocator.  For the main
+    // interpreter and for all interpreters that don't have their
+    // own obmalloc state, this points to the static structure in
+    // obmalloc.c obmalloc_state_main.  For other interpreters, it is
+    // heap allocated by _PyMem_init_obmalloc() and freed when the
+    // interpreter structure is freed.  In the case of a heap allocated
+    // obmalloc state, it is not safe to hold on to or use memory after
+    // the interpreter is freed. The obmalloc state corresponding to
+    // that allocated memory is gone.  See free_obmalloc_arenas() for
+    // more comments.
+    struct _obmalloc_state *obmalloc;
```
There are a couple places where things are simpler if we did not make this a pointer.  That only works, though, if we statically initialize `obmalloc_state_main` and keep the static initializer code in pycore_obmalloc_init.h and pycore_runtime_init.h.

On the other handle we'd be wasting a little bit of memory for the common case (main interpreter).

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -639,6 +643,13 @@ pycore_create_interpreter(_PyRuntimeState *runtime,
+    // any calls are made to obmalloc functions.
+    if (!_PyMem_init_obmalloc(interp, runtime)) {
+        return  _PyStatus_NO_MEMORY();
+    }
```
If we statically initialized `obmalloc_state_main` then we wouldn't need to do this, would we?

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -2121,6 +2132,14 @@ new_interpreter(PyThreadState **tstate_p, const PyInterpreterConfig *config)
+    if (!_PyMem_init_obmalloc(interp, runtime)) {
+        status = _PyStatus_NO_MEMORY();
+        goto error;
+    }
```
Likewise, if we did not make `PyInterpreterState.obmalloc` a pointer and kept the existing static initializer code, then we wouldn't need to do this.

ericsnowcurrently replied:
Then again, it's basically the same as what we were doing in `init_interpreter()`.  At the same time, it seemed a bit more natural where it was in `init_interpreter()` and seems more out of place here.  Note that we've been making some effort over the last few years to make the different steps in initialization less coupled and more compartmentalized.

You're right about needing the config though.  Perhaps we should be passing the config to `PyInterpreterState_New()` and move the `init_interp_settings()` and `_PyMem_init_obmalloc()` calls down.

Regardless, if there isn't a simple obvious solution on this we can punt to later.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -547,6 +547,11 @@ free_interpreter(PyInterpreterState *interp)
+            // interpreter has its own obmalloc state, free it
+            PyMem_RawFree(interp->obmalloc);
+            interp->obmalloc = NULL;
+        }
```
This could also go away if we didn't make `PyInterpreterState.obmalloc` a pointer.

ericsnowcurrently reviewed (changes requested):
Thanks for doing this!  I think this is basically right.  I didn't see any correctness issues.

Clearly I did leave a bunch of comments, but it's mostly little things like alternate spellings, whether or not to statically initialize `obmalloc_state_main`, and whether or not we really have to make `PyInterpreterState.obmalloc` a pointer.

nascheme committed:
Remove redundant init of .obmalloc member

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

nascheme committed:
Code cleanup based on review feedback.

- Remove 'runtime' argument.
- Remove 'initialized' and 'heap_allocated' members.
- Use _Py_IsMainInterpreter()
- Remove spurious whitespace change.
- Return -1 on error and 0 on success, as normal convention.

ericsnowcurrently commented on Tools/c-analyzer/cpython/ignored.tsv:
```diff
@@ -325,7 +325,7 @@ Objects/obmalloc.c	-	_PyMem_Debug	-
 Objects/obmalloc.c	-	_PyObject	-
 Objects/obmalloc.c	-	last_final_leaks	-
-Objects/obmalloc.c	-	usedpools	-
+Objects/obmalloc.c	-	obmalloc_state_main	-
```
```suggestion
Objects/obmalloc.c	-	obmalloc_state_main	-
Objects/obmalloc.c	-	obmalloc_state_initialized	-
```

ericsnowcurrently reviewed (approved):
LGTM

nascheme committed:
Update c-analyzer globals list.

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>
